File: implementations/swift/Sources/RCLParser/Formatter.swift
import Foundation

public enum Formatter {
    public static func format(_ doc: DocumentNode) -> String {
        if let root = doc.rootValue, root.kind == "array" { return "do \(formatValue(root))" }
        doc.blocks.map { formatBlock($0, indent: 0) }.joined(separator: "\n\n")
    }

    private static func formatBlock(_ block: BlockNode, indent: Int) -> String {
        let pad = String(repeating: "  ", count: indent)
        let head = block.argument != nil
            ? "\(pad)\(block.name) \(q(block.argument!)) do"
            : "\(pad)\(block.name) do"
        var lines = [head]

        for (k, v) in block.properties {
            if v.kind == "array" { lines.append("\(pad)  \(k) do \(formatValue(v)) end") }
            else { lines.append("\(pad)  \(k) = \(formatValue(v))") }
        }

        var seen = Set<String>()
        for child in block.blocks.values {
            let key = child.argument != nil ? "\(child.name):\(child.argument!)" : child.name
            if seen.insert(key).inserted { lines.append(formatBlock(child, indent: indent + 1)) }
        }
        for child in block.namedBlocks {
            let key = child.argument != nil ? "\(child.name):\(child.argument!)" : child.name
            if seen.insert(key).inserted { lines.append(formatBlock(child, indent: indent + 1)) }
        }

        lines.append("\(pad)end")
        return lines.joined(separator: "\n")
    }

    private static func formatValue(_ node: Node) -> String {
        switch node {
        case .string(let n): return q(n.value)
        case .number(let n): return "\(n.value)"
        case .boolean(let n): return n.value ? "true" : "false"
        case .array(let n): return "[\(n.elements.map { formatValue($0) }.joined(separator: ", "))]"
        case .block(let n): return formatAnonymousBlock(n)
        }
    }

    private static func formatAnonymousBlock(_ block: BlockNode) -> String {
        var parts: [String] = []
        for (k, v) in block.properties { parts.append("\(k) = \(formatValue(v))") }
        for child in block.blocks.values { parts.append(formatInlineBlock(child)) }
        for child in block.namedBlocks { parts.append(formatInlineBlock(child)) }
        return parts.isEmpty ? "do end" : "do \(parts.joined(separator: " ")) end"
    }

    private static func formatInlineBlock(_ block: BlockNode) -> String {
        let head = block.argument != nil ? "\(block.name) \(q(block.argument!)) do" : "\(block.name) do"
        var parts: [String] = []
        for (k, v) in block.properties { parts.append("\(k) = \(formatValue(v))") }
        for child in block.blocks.values { parts.append(formatInlineBlock(child)) }
        for child in block.namedBlocks { parts.append(formatInlineBlock(child)) }
        return parts.isEmpty ? "\(head) end" : "\(head) \(parts.joined(separator: " ")) end"
    }

    private static func q(_ value: String) -> String {
        let escaped = value
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\"", with: "\\\"")
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: "\t", with: "\\t")
        return "\"\(escaped)\""
    }
}

File: implementations/ruby/lib/rcl/formatter.rb
module RCL
  class Formatter
    def format(ast)
      if ast["root_value"]&.fetch("kind", nil) == "array"
        return "do #{format_value(ast.fetch("root_value"))}"
      end
      ast.fetch("blocks").map { |b| format_block(b, 0) }.join("\n\n")
    end

    private

    def format_block(block, indent)
      pad = "  " * indent
      header = if block["argument"]
                 "#{pad}#{block.fetch("name")} #{quote(block.fetch("argument"))} do"
               else
                 "#{pad}#{block.fetch("name")} do"
               end
      lines = [header]

      block.fetch("properties", {}).each do |k, v|
        if v.fetch("kind") == "array"
          lines << "#{pad}  #{k} do #{format_value(v)} end"
        else
          lines << "#{pad}  #{k} = #{format_value(v)}"
        end
      end

      block.fetch("blocks", {}).each_value do |child|
        lines << format_block(child, indent + 1)
      end
      block.fetch("named_blocks", []).each do |child|
        lines << format_block(child, indent + 1)
      end

      lines << "#{pad}end"
      lines.join("\n")
    end

    def format_value(node)
      case node.fetch("kind")
      when "string"
        quote(node.fetch("value"))
      when "number"
        node.fetch("value").to_s
      when "boolean"
        node.fetch("value") ? "true" : "false"
      when "array"
        "[#{node.fetch("elements").map { |e| format_value(e) }.join(", ")}]"
      when "block"
        format_anonymous_block(node)
      else
        raise "unsupported node kind #{node.fetch("kind")}"
      end
    end

    def format_anonymous_block(block)
      parts = []
      block.fetch("properties", {}).each { |k, v| parts << "#{k} = #{format_value(v)}" }
      block.fetch("blocks", {}).each_value { |child| parts << format_inline_block(child) }
      block.fetch("named_blocks", []).each { |child| parts << format_inline_block(child) }
      parts.empty? ? "do end" : "do #{parts.join(" ")} end"
    end

    def format_inline_block(block)
      head = if block["argument"]
               "#{block.fetch("name")} #{quote(block.fetch("argument"))} do"
             else
               "#{block.fetch("name")} do"
             end
      parts = []
      block.fetch("properties", {}).each { |k, v| parts << "#{k} = #{format_value(v)}" }
      block.fetch("blocks", {}).each_value { |child| parts << format_inline_block(child) }
      block.fetch("named_blocks", []).each { |child| parts << format_inline_block(child) }
      parts.empty? ? "#{head} end" : "#{head} #{parts.join(" ")} end"
    end

    def quote(value)
      escaped = value.gsub("\\", "\\\\").gsub("\"", "\\\"").gsub("\n", "\\n").gsub("\t", "\\t")
      "\"#{escaped}\""
    end
  end
end
